### PR TITLE
navigator: add enum types

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR continue the effort to add `Enum` type to the SDK (See issue #105)

This PR add `Enum` type support for the navigator.  

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: navigator/backend/src/main/scala/com/digitalasset/navigator/model/converter/LedgerApiV1.scala [hunk 1]
```diff
@@ -372,6 +372,30 @@ case object LedgerApiV1 {
...
+        .toRight(GenericConversionError(s"Unknown type ${typeCon.name.identifier}"))
+      dt <- typeCon.instantiate(ddt) match {
+        case v @ iface.Enum(_) => Right(v)
+        case iface.Record(_) | iface.Variant(_) =>
+          Left(GenericConversionError(s"Enum expected"))
+      }
+      _ <- Either.cond(
+        dt.constructors.contains(enum.constructor),
+        (),
+        GenericConversionError(s"Unknown choice ${enum.constructor}"))
```
Comment (rautenrieth-da) [inline navigator/backend/src/main/scala/com/digitalasset/navigator/model/converter/LedgerApiV1.scala]:
```suggestion
        GenericConversionError(s"Unknown enum constructor ${enum.constructor}"))
```

File: navigator/frontend/src/applets/templatecontracts/columns.tsx [hunk 2]
```diff
@@ -1,26 +1,21 @@
...
 // SPDX-License-Identifier: Apache-2.0
 
-import {
-  ChoicesButton,
-  ContractColumn,
-} from '@da/ui-core';
-import { DamlLfDataType, DamlLFFieldWithType } from '@da/ui-core/lib/api/DamlLfType';
-import { DamlLfValueRecord } from '@da/ui-core/lib/api/DamlLfValue';
+import {ChoicesButton, ContractColumn} from '@da/ui-core';
+import {DamlLFFieldWithType, DamlLfRecord} from '@da/ui-core/lib/api/DamlLfType';
```
Comment (rautenrieth-da) [inline navigator/frontend/src/applets/templatecontracts/columns.tsx]: Merge this line with the `import {DamlLfValueRecord} from '@da/ui-core/lib/api/DamlLfValue';` line below

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx [hunk 3]
```diff
@@ -2,9 +2,10 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import * as React from 'react';
+import * as DamlLfTypeF from '../api/DamlLfType';
 import {
   DamlLfDataType,
-  DamlLfDefDataType,
+  DamlLfDefDataType, DamlLfEnum,
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: Newline between those two imports

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx [hunk 4]
```diff
@@ -13,31 +14,33 @@ import {
   DamlLfTypePrim,
   DamlLfVariant,
 } from '../api/DamlLfType';
-import * as DamlLfTypeF from '../api/DamlLfType';
 import {
   DamlLfValue,
   DamlLfValueBool,
-  DamlLfValueDecimal,
+  DamlLfValueDecimal, DamlLfValueEnum,
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: Newline between those two imports

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx [hunk 5]
```diff
@@ -13,31 +14,33 @@ import {
...
-  DamlLfValueList, DamlLfValueMap,
+  DamlLfValueList,
+  DamlLfValueMap,
   DamlLfValueOptional,
   DamlLfValueParty,
   DamlLfValueRecord,
   DamlLfValueText,
   DamlLfValueUnit,
-  DamlLfValueVariant, mapEntry,
+  DamlLfValueVariant, enumCon,
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: Newline between those two imports

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx [hunk 6]
```diff
@@ -694,9 +737,20 @@ class TypeConInput extends React.Component<TypeConInputProps, TypeConInputState>
     const { parameter, onChange } = this.props;
     if (ddt) {
       const dataType = DamlLfTypeF.instantiate(parameter, ddt);
-      const initialValue = DamlLfValueF.initialDataTypeValue(parameter.name, dataType);
+      console.log('TUKA ' + ddt.dataType.type)
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: Remove this line

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx [hunk 7]
```diff
@@ -728,7 +782,8 @@ class TypeConInput extends React.Component<TypeConInputProps, TypeConInputState>
     const { dataType } = this.state;
 
     if (dataType === undefined) {
-      return <em>Loading type {parameter.name.name}...</em>;
+      console.log('RoDer ' + parameter.name.name );
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: Remove this line

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx [hunk 8]
```diff
@@ -728,7 +782,8 @@ class TypeConInput extends React.Component<TypeConInputProps, TypeConInputState>
...
+      return <em>JCVD Loading type {parameter.name.name}...</em>;
```
Comment (rautenrieth-da) [inline navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: What is JCVD? [Jean-Claude Van Damme](https://www.google.com/search?q=jcvd)?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/CommandPreprocessor.scala [hunk 9]
```diff
@@ -243,6 +244,33 @@ private[engine] class CommandPreprocessor(compiledPackages: ConcurrentCompiledPa
                 }
             }
 
+          case (TTyCon(id), ValueEnum(mbId, constructor)) =>
+            mbId match {
+              case Some(id_) if id_ != id =>
+                fail(s"Mismatching enum id, the types tell us $id, but the value tells us $id_")
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/CommandPreprocessor.scala]: changes 3 times

File: navigator/backend/src/main/scala/com/digitalasset/navigator/GraphQLHandler.scala [hunk 10]
```diff
@@ -72,7 +72,7 @@ case class DefaultGraphQLHandler(
       Future.successful(InternalServerError -> JsString("Platform store not available"))
     )(store => {
       val context = GraphQLContext(party, store)
-      Executor
+      val x1 = Executor
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/GraphQLHandler.scala]: fixed

File: navigator/backend/src/main/scala/com/digitalasset/navigator/model/converter/LedgerApiV1.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/model/converter/LedgerApiV1.scala]: fixed

File: navigator/frontend/src/applets/templatecontracts/columns.tsx (hunk 2 above)
Comment (remyhaemmerle-da) [reply navigator/frontend/src/applets/templatecontracts/columns.tsx]: fixed

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx (hunk 4 above)
Comment (remyhaemmerle-da) [reply navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: fixed

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx (hunk 5 above)
Comment (remyhaemmerle-da) [reply navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: fixed

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx (hunk 6 above)
Comment (remyhaemmerle-da) [reply navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: fixed

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx (hunk 7 above)
Comment (remyhaemmerle-da) [reply navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: fixed

File: navigator/frontend/src/ui-core/src/ParameterForm/index.tsx (hunk 8 above)
Comment (remyhaemmerle-da) [reply navigator/frontend/src/ui-core/src/ParameterForm/index.tsx]: :S

removed

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala [hunk 11]
```diff
@@ -57,6 +57,9 @@ object InterfaceReader {
     def addVariant(k: QualifiedName, tyVars: ImmArraySeq[Ref.Name], a: Variant.FWT): State =
       this.copy(typeDecls = this.typeDecls.updated(k, InterfaceType.Normal(DefDataType(tyVars, a))))
 
+    def addEnum[E](k: QualifiedName, tyVars: ImmArraySeq[Ref.Name], a: Enum) =
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala]:
```suggestion
    def addEnum(k: QualifiedName, tyVars: ImmArraySeq[Ref.Name], a: Enum) =
```

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala [hunk 12]
```diff
@@ -266,6 +288,9 @@ object InterfaceReader {
       ctx: Context): InterfaceReaderError \/ FieldWithType =
     type_(a.getType, ctx).flatMap(t => name(a.getField).map(_ -> t))
 
+  private def constructors(as: DamlLf1.DefDataType): InterfaceReaderError \/ ImmArray[Name] =
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala]: Variants have constructors too.

```suggestion
  private def enumConstructors(as: DamlLf1.DefDataType): InterfaceReaderError \/ ImmArraySeq[Name] =
```

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala [hunk 13]
```diff
@@ -170,6 +175,23 @@ object InterfaceReader {
...
+      case (st, (k, typVars, enum)) => st.addEnum(k, typVars, enum)
+    }
+
+  private[reader] def enum(m: ModuleName)(a: DamlLf1.DefDataType)
+    : InterfaceReaderError.Tree \/ (QualifiedName, ImmArraySeq[Ref.Name], Enum) =
+    (locate('name, rootErrOf[ErrorLoc](fullName(m, a.getName))).validation |@|
+      locate('typeParams, typeParams(a)).validation |@|
+      locate('constructors, rootErrOf[ErrorLoc](constructors(a))).validation) {
+      (k, tyVars, constructors) =>
+        (k, tyVars.toSeq, Enum(constructors.toSeq))
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala]: `(_:ImmArraySeq[...]).toImmArray.toSeq` = `identity`

```suggestion
        (k, tyVars.toSeq, Enum(constructors))
```

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala [hunk 14]
```diff
@@ -66,8 +66,8 @@ object DataType {
             Traverse[Record].traverse(r)(f).widen
           case v @ Variant(_) =>
             Traverse[Variant].traverse(v)(g).widen
-          case Enum(vs) =>
-            Applicative[G].pure(Enum(vs))
+          case e: Enum =>
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala]:
```suggestion
          case e @ Enum(_) =>
```

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala (hunk 11 above)
Comment (remyhaemmerle-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala]: fixed

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala (hunk 12 above)
Comment (remyhaemmerle-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala]: done

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala (hunk 13 above)
Comment (remyhaemmerle-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/reader/InterfaceReader.scala]: done

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala (hunk 14 above)
Comment (remyhaemmerle-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala]: done

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueVersion.scala [hunk 15]
```diff
@@ -52,8 +51,8 @@ object ValueVersions
               case ValueMap(map) =>
                 go(maxVV(minMap, currentVersion), map.values ++: values)
               case ValueEnum(_, _) =>
-                go(maxVV(minEnum, currentVersion), values)
-              // tuples are a no-no
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueVersion.scala]: Don't delete this comment.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueVersion.scala]: fixed

**Pull request merged**